platform/engineering/request-for-change/2019-09-20-rfc-cms-based-alerts.md

```diff
@@ -0,0 +1,84 @@
+
+This functionality does not currently exist. Each of the alert implementations discussed above has significant limitations in attempting to satisfy the story:
+
+1. **Existing BRD process**. This takes more than ten minutes to run and rebuilds the entire site. Feels very weighty for what may regularly be a small update on a single page.
+1. **Teamsite**. This is going away once legacy VA.gov pages are migrated to WBC VA.gov or retired.
+1. **Sourced from `va.gov-content` on load**. Requires content managers to be comfortable with GitHub and markdown or YAML.
+1. **Sourced from Vets API on load.** No logical connection between the CMS and Vets API exists.
+
+## Possible design approaches
+
+For the first two possibilities, assume that a change to a specific area of the content model causes the CMS to tell Jenkins kick off a specific pipeline with certain parameters. For the third, Jenkins is out of the loop.
+
+### Single page BRD
+
+Jenkins makes a GraphQL query the CMS content API based on given parameters, pulls _only_ the relevant content and flags, builds _only_ the corresponding templates, runs _only_ the relevant tests and checks, and uploads the HTML asset to the existing S3 bucket, overwriting the existing asset. As soon as upload is complete, the page begins being served to users. Alerts are rendered along with the rest of the static content on the page.
+
+|Pros|Cons|
+|----|----|
+|Familiar architecture and process|Somewhat weighty process for minor updates|
+|Can apply existing tests (e.g. link checker, 508)|Inherently slower b/c of tests and checks|
+|No additional client requests needed after page download|
+|Leverages benefits of GraphQL|
+
+### Sourced from S3 bucket on load
+
+Jenkins makes a GraphQL query to the CMS content API based on given parameters, pulls _only_ the relevant content and flags, builds one or more JSON, YAML, or other type of structured data file, and uploads to a dedicated S3 bucket, possibly overwriting the existing asset. On page load, a React component makes an asynchronous call to the S3 bucket, parses the payload for any relevant alert information, and renders the result (if applicable).
+
+|Pros|Cons|
+|----|----|
+|Lightweight and fast|Need to track another kind of asset outside of source control|
+|Leverages benefits of GraphQL|Adds more architectural and process complexity website management|
+| |Requires at least one more client request after page download|
+
+### Sourced from CMS content API on load
+
+A React component makes an asynchronous call with page-specific parameters to api.va.gov (or some other middleware that shares network or environment access with the CMS content API) which, in turn, makes a GraphQL call to the CMS content API for _only_ the relevant content and flags. The payload is returned to the React component which renders the result (if applicable).
```

Comment:
The data from the CMS could be cached in the database for the API. This would be similar to the way we do maintenance windows from PagerDuty.

- We have a sidekiq (background) job that queries for maintenance windows from the PagerDuty API, stores them in the database.
- A react component makes a call to the API to load relevant data (does NOT pass through to the Drupal backend)

There's no reason that this would have to be done in realtime as described here unless i'm missing something.

Reply:
Perhaps ^^ should be a separate potential implementation?

Reply:
This is the first time the vets-api in production would talk to the CMS.  May need to update the ATO to acknowledge a new system integration and new data stored in the API (even though it is "cached" only meaning it is transitive and not a permanent source of truth or contain PII.